src/Bicep.Core/Semantics/Namespaces/NamespaceResolver.cs

```diff
@@ -16,62 +16,37 @@ namespace Bicep.Core.Semantics.Namespaces
     public class NamespaceResolver
     {
         private readonly ImmutableDictionary<string, NamespaceType> namespaceTypes;
+        public ImmutableDictionary<string, BuiltInNamespaceSymbol> BuiltIns { get; }
 
-        private NamespaceResolver(ImmutableDictionary<string, NamespaceType> namespaceTypes, ImmutableDictionary<string, BuiltInNamespaceSymbol> builtIns)
-        {
-            this.namespaceTypes = namespaceTypes;
-            this.BuiltIns = builtIns;
-        }
-
-        public static NamespaceResolver Create(IFeatureProvider features, INamespaceProvider namespaceProvider, BicepSourceFile sourceFile, ResourceScope targetScope, ILanguageScope fileScope)
+        public static NamespaceResolver Create(ImmutableArray<NamespaceResult> namespaceResults)
         {
-            var declaredNamespaces = fileScope.Declarations.OfType<ProviderNamespaceSymbol>()
-                .DistinctBy(x => x.Name, LanguageConstants.IdentifierComparer);
-
-            var builtInNamespaceSymbols = new Dictionary<string, BuiltInNamespaceSymbol>(LanguageConstants.IdentifierComparer);
-
-            var namespaceTypes = declaredNamespaces
-                .Select(x => x.DeclaredType)
-                .OfType<NamespaceType>()
-                .ToImmutableDictionary(x => x.Name, LanguageConstants.IdentifierComparer);
+            Dictionary<string, NamespaceType> namespaceTypes = new(LanguageConstants.IdentifierComparer);
+            Dictionary<string, BuiltInNamespaceSymbol> builtIns = new(LanguageConstants.IdentifierComparer);
```

Comment:
It might be better to use ImmutableDictionary.CreateBuilder to avoid some overhead in building Dictionary.